File: misc/FoldersAndCows/foldersAndCows.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

class Folder{
	int id = 0; 
	boolean access = false; // true: shared, false: confidential
	List<Folder> children = new LinkedList<>(); 
	List<Integer> cows = new LinkedList<>();
	int in = 0;
	int out = 0;

	Folder(int id, boolean access){
		this.id = id;
		this.access = access;
		in = 0;
		out = 0;
	}

	void appendCow(List<Integer> newCows){
		cows.addAll(newCows);
	}
}

public class Solution {
    private static Folder findFolder(List<Folder> folders, int id){
		for(Folder folder: folders)
			if(folder.id == id)
				return folder;
		return null;
	}
    
    private static void traverse(int[] cows, Folder root){
		if(root.out == 0)
			for(int i : root.cows)
				cows[i]++;
		for(Folder f: root.children){
			if(f.access)
				f.appendCow(root.cows);
			traverse(cows, f);
		}
	}
    
    public static void main(String args[] ) throws Exception {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT */
        Scanner scanner = new Scanner(System.in);
		List<Folder> folders = new ArrayList<>();


		// parse and build folders
		int Q = scanner.nextInt();
		int M = scanner.nextInt();
		int N = scanner.nextInt();
		while(M-- > 0){
			Folder f = new Folder(scanner.nextInt(), true);
			folders.add(f);
			int K = scanner.nextInt();
			while(K-- > 0)
				f.cows.add(scanner.nextInt());
		}
		while(N-- > 0){
			Folder f = new Folder(scanner.nextInt(), false);
			folders.add(f);
			int K = scanner.nextInt();
			while(K-- > 0)
				f.cows.add(scanner.nextInt());
		}

		// parse and build trees
		int G = scanner.nextInt();
		while(G-- > 0){
			int U = scanner.nextInt();
			int V = scanner.nextInt();
			Folder u = findFolder(folders, U);
			Folder v = findFolder(folders, V);
			if(u != null & v != null){
				u.children.add(v);
				u.out++;
				v.in++;
			}
		}

		// traverse
		int numLeaf = 0;
		int[] cows = new int[Q];
		for(int i = 0; i < folders.size(); i++){
			if(folders.get(i).in == 0)
				traverse(cows, folders.get(i));
			if(folders.get(i).out == 0)
				numLeaf++;
		}

		List<Integer> res = new ArrayList<>();
		for(int i = 0; i < Q; i++)
			if(cows[i] < numLeaf)
				res.add(i);

		for(int i = 0; i < res.size(); i++)
			System.out.print(res.get(i) + " ");
	}
}

File: misc/FoldersAndCows/foldersAndCows.py
import collections
import fileinput

class Folder:
    def __init__(self, id, shared):
        self.id = id
        self.shared = shared # boolean
        # Number of incoming arcs
        self.incoming = 0
        # Number of outcoming arcs
        self.outcoming = 0
        self.cows = []
        self.children = []

    def appendCow(self, newCows):
        self.cows.extend(newCows)

def findFolder(folders, id):
    try:
        return folders[id]
    except KeyError:
        return None

def traverse(cows, root):
    if root.outcoming == 0:
        # There is no outcoming arcs from the root
        # Thus, we are at a leaf node
        for i in root.cows:
            cows[i] += 1
    for f in root.children:
        if f.shared:
            f.appendCow(root.cows)
        traverse(cows, f)

if __name__ == "__main__":
    # HackerRank way to handle the stdin input
    # The advantage of doing is that we can have a unified interface to
    # handle both file input and stdin input. File input is easier to debug locally
    # and we don't need to modify the code to allow the code to run on HackerRank using stdin input.
    # We can do this because of fileinput module: supporting two ways of passing at the same time:
    # - support foldersAndCows.py < input000.txt
    # - support foldersAndCows.py input000.txt
    input_string = ""
    for line in fileinput.input():
        input_string += line
    input_string_list = collections.deque(input_string.split('\n'))

    # the number of cows
    Q = int(input_string_list.popleft())
    # M: the number of shared folder; N: the number of confidential folder
    M,N = list(map(int, input_string_list.popleft().split(' ')))
    folders = {}

    while M > 0:
        line = list(map(int, input_string_list.popleft().split(' ')))
        f = Folder(line[0], True)
        folders[line[0]] = f
        k = line[1]
        if k != 0:
            f.cows.extend(list(line[2:]))
        M -= 1

    while N > 0:
        line = list(map(int, input_string_list.popleft().split(' ')))
        f = Folder(line[0], False)
        folders[line[0]] = f
        k = line[1]
        if k != 0:
            f.cows.extend(list(line[2:]))
        N -= 1

    # parse and build trees
    G = int(input_string_list.popleft())
    while G > 0:
        G -= 1
        U,V = list(map(int, input_string_list.popleft().split(' ')))
        u = findFolder(folders, U)
        v = findFolder(folders, V)
        if u and v:
            u.children.append(v)
            u.outcoming += 1
            v.incoming += 1

    # traverse
    numLeaf = 0
    cows = [0] * Q
    for i in folders.keys():
        if folders[i].incoming == 0:
            traverse(cows, folders[i])
        if folders[i].outcoming == 0:
            numLeaf += 1

    res = []
    for i in range(Q):
        if cows[i] < numLeaf:
            res.append(str(i))

    print(' '.join(res))
